firedrake/preconditioners/hypre_ams.py

```diff
@@ -0,0 +1,69 @@
+from firedrake.preconditioners.base import PCBase
+from firedrake.petsc import PETSc
+from firedrake.functionspace import FunctionSpace
+from firedrake.constant import Constant
+from firedrake.ufl_expr import TestFunction
+from firedrake.interpolation import Interpolator
+from firedrake.projection import project
+from firedrake.dmhooks import get_function_space
+from ufl import grad
+
+__all__ = ("HypreAMS",)
+
+
+class HypreAMS(PCBase):
+    def initialize(self, obj):
```

Comment:
It would be good to add a bibtex reference to the relevant paper describing the hypre implementation. I think it is this one? https://hypre.readthedocs.io/en/latest/ch-references.html#kova2009

To do that, you do:

```
from firedrake_citations import Citations

Citations().add("Kolev2009", """
BIBTEX STRING
""")
```

Then inside `initialize` do:


```
Citations().register("Kolev2009")
```

Thanks

Reply:
After adding the citation another strange CI error occured. Now PETSc reports that the PC type `hypre` is not known. Did anything else change? Maybe `register` should be moved to `__init__` instead of `initialize`?

Reply:
Ah, I see. The problem seems to occur in the complex test, where hypre is unavailable.

Reply:
Add `@pytest.mark.skipcomplex` as a decorator to the two tests.